lib/collection.js

```diff
@@ -0,0 +1,290 @@
+var Events = require('./events'),
+    Collection,
+    ARRAY_PROXY_METHODS = ['forEach', 'some', 'every', 'filter', 'map', 'reduce'];
+
+function clone(obj) {
+    return Object.keys(obj).reduce(function (result, key) {
+        result[key] = obj[key];
+        return result;
+    }, {});
+}
+
+/**
+ * @class Collection
+ * @extends Events
+ */
+Collection = Events.inherit(/** @lends Collection.prototype*/{
+
+    /**
+     * @type {Function}
+     */
+    modelType: null,
+
+    /**
+     * @param {Array.<(Model|Object)>} [data]
+     */
+    __constructor: function (data) {
+        this.__base.apply(this, arguments);
+
+        /**
+         * @type {Array.<Model>}
+         */
+        this._models = [];
+
+        /**
+         * @type {Object.<string, Model>}
+         */
+        this._modelsIdsMap = {};
+
+        Object.defineProperty(this, 'length', {
+            get: function () {
+                return this._models.length;
+            }
+        });
+
+        data = data || [];
+        this.set(data);
+    },
+
+    /**
+     * @param {String} [branch=DEFAULT_BRANCH]
+     * @returns {Boolean}
+     */
+    isChanged: function (branch) {
+        return this._models.some(function (model) {
+            return model.isChanged(branch);
+        });
```

Comment:
Why lists? you want to avoid isChanged = true when model is extracted from collection and then inserted in same place?
Also should collection be changed when models are just reordered?